core/unix/signal.c

```diff
@@ -4801,7 +4801,15 @@ master_signal_handler_C(byte *xsp)
         return;
     }
 #endif
-    dcontext_t *dcontext = get_thread_private_dcontext();
+    /* We avoid using safe_read_tls_magic during detach. This thread may already have
+     * lost its TLS. A safe read may result into a race affecting asynchronous non-alarm
+     * signals (xref i#3535) between delivering the SIGSEGV and restoring the app's signal
+     * handlers. We don't need the thread's private dcontext anymore here at this point.
+     * Note that there is still a small race window if the signal gets delivered after the
+     * detach has finished, i.e. doing_detach is false. This is an issue in particular if
+     * the app has started re-attaching.
+     */
+    dcontext_t *dcontext = doing_detach ? NULL : get_thread_private_dcontext();
```

Comment:
I think this needs to explain that there is special logic to avoid the thread_lookup below: I found it confusing to read b/c I kept thinking "but it's just going to do a lookup and not stay NULL: so is the point just to avoid the safe read, and really it's not to stay NULL".

Reply:
Does it need to avoid the thread_lookup?

Reply:
I see. I think the thread_lookup always returns null, since at this point, the thread is already running native (resumed after synch in detach), and has exited, i.e. removed from the hash.

Reply:
True. Correction: "thread *may* be already running natively". So there are two separate windows. 1) doing_detach, dcontext still valid: it will look up the thread and find it anyway. 2) doing_detach, dcontext destroyed: lookup will fail.

Reply:
1) not true. We did not do a thread_lookup. In any case, patch was changed into dropping before anything happens.